src/clp_ffi_js/ir/StructuredIrStreamReader.hpp

```diff
@@ -0,0 +1,201 @@
+              m_log_level_key{std::move(log_level_key)},
+              m_timestamp_key{std::move(timestamp_key)} {}
+
+    /**
+     * Handles a log event by inserting it into the deserialized log events vector.
+     * @param log_event
+     * @return IRErrorCode::IRErrorCode_Success
+     */
+    [[nodiscard]] auto handle_log_event(clp::ffi::KeyValuePairLogEvent&& log_event
+    ) -> clp::ffi::ir_stream::IRErrorCode {
+        m_deserialized_log_events->emplace_back(std::move(log_event));
+
+        return clp::ffi::ir_stream::IRErrorCode::IRErrorCode_Success;
+    }
+
+    /**
+     * @param utc_offset_old
+     * @param utc_offset_new
+     * @return IRErrorCode::IRErrorCode_Success
+     */
+    [[nodiscard]] static auto handle_utc_offset_change(
+            [[maybe_unused]] clp::UtcOffset utc_offset_old,
+            [[maybe_unused]] clp::UtcOffset utc_offset_new
+    ) -> clp::ffi::ir_stream::IRErrorCode {
+        SPDLOG_WARN("UTC offset change packets are currently not handled.");
+
+        return clp::ffi::ir_stream::IRErrorCode::IRErrorCode_Success;
+    }
+
+    /**
+     * Handles the insertion of a schema tree node by finding node IDs for log level and
+     * timestamp keys.
+     * @param schema_tree_node_locator
+     * @return IRErrorCode::IRErrorCode_Success
+     */
+    [[nodiscard]] auto handle_schema_tree_node_insertion(
+            clp::ffi::SchemaTree::NodeLocator schema_tree_node_locator
+    ) -> clp::ffi::ir_stream::IRErrorCode {
+        ++m_current_node_id;
+
+        auto const& key_name{schema_tree_node_locator.get_key_name()};
+        if (m_log_level_key == key_name) {
+            m_level_node_id.emplace(m_current_node_id);
+        } else if (m_timestamp_key == key_name) {
+            m_timestamp_node_id.emplace(m_current_node_id);
+        }
+
+        return clp::ffi::ir_stream::IRErrorCode::IRErrorCode_Success;
+    }
+
+    /**
+     * @return IRErrorCode::IRErrorCode_Success
+     */
+    [[nodiscard]] static auto handle_end_of_stream() -> clp::ffi::ir_stream::IRErrorCode {
+        return clp::ffi::ir_stream::IRErrorCode::IRErrorCode_Success;
+    }
+
+    // Methods
+    /**
+     * @return The node ID associated with the log level key.
+     */
+    [[nodiscard]] auto get_level_node_id() const -> parsed_tree_node_id_t {
+        return m_level_node_id;
+    }
+
+    /**
+     * @return The node ID associated with the timestamp key.
+     */
+    [[nodiscard]] auto get_timestamp_node_id() const -> parsed_tree_node_id_t {
+        return m_timestamp_node_id;
+    }
+
+private:
+    // Variables
+    std::string m_log_level_key;
+    std::string m_timestamp_key;
+
+    clp::ffi::SchemaTree::Node::id_t m_current_node_id{clp::ffi::SchemaTree::cRootId};
+
+    parsed_tree_node_id_t m_level_node_id;
+    parsed_tree_node_id_t m_timestamp_node_id;
+    std::shared_ptr<std::vector<clp::ffi::KeyValuePairLogEvent>> m_deserialized_log_events;
```

Comment:
Might be a typo: should it be gsl::not_null?

Reply:
Yes, sorry.

Reply:
np. updated as below
```suggestion

    // TODO: Technically, we don't need to use a `shared_ptr` since the parent stream reader will
    // have a longer lifetime than this class. Instead, we could use `gsl::not_null` once we add
    // `gsl` into the project.
    std::shared_ptr<std::vector<clp::ffi::KeyValuePairLogEvent>> m_deserialized_log_events;
```